muxfs/mux.go

```diff
@@ -31,53 +31,52 @@ type Mux struct {
 var _ qfs.Filesystem = (*Mux)(nil)
 
 // SetHandler designates the resolver for a given path kind string
-func (m *Mux) SetHandler(pathKind string, resolver qfs.Filesystem) {
+func (m *Mux) SetFilesystem(fs qfs.Filesystem) error {
 	if m.handlers == nil {
 		m.handlers = map[string]qfs.Filesystem{}
 	}
-	m.handlers[pathKind] = resolver
-}
 
-// GetHandler returns the resolver for a given path kind string and a bool
-// if the resolver exists on the muxfs
-func (m *Mux) GetHandler(pathKind string) (qfs.Filesystem, bool) {
-	resolver, ok := m.handlers[pathKind]
-	return resolver, ok
-}
+	if m.handlers[fs.Type()] != nil {
+		return fmt.Errorf("mux already has a %q filesystem", fs.Type())
+	}
+
+	if releaser, ok := fs.(qfs.ReleasingFilesystem); ok {
+		m.doneWg.Add(1)
+		go func(releaser qfs.ReleasingFilesystem) {
+			<-releaser.Done()
+			m.doneErr = releaser.DoneErr()
+			m.doneWg.Done()
+		}(releaser)
+	}
+	if m.defaultWriteDestination == "" {
+		if _, ok := fs.(cafs.Filestore); ok {
+			m.defaultWriteDestination = fs.Type()
+		}
+	}
 
-// Option is a function that manipulates config details when fed to New(). Fields on
-// the o parameter may be null, functions cannot assume the Config is non-null.
-type Option func(o *[]MuxConfig) error
+	m.handlers[fs.Type()] = fs
+	return nil
+}
 
-// MuxConfig contains the information needed to create a new filesystem
-type MuxConfig struct {
-	Type   string                 `json:"type"`
-	Config map[string]interface{} `json:"config,omitempty"`
-	Source string                 `json:"source,omitempty"`
+// Filesystem returns the filesystem for a given fs type string and a bool
+// if the resolver exists on the muxfs
+func (m *Mux) Filesystem(fsType string) qfs.Filesystem {
+	return m.handlers[fsType]
 }
 
 // constructors maps filesystem type strings to constructor functions
-var constructors = map[string]qfs.FSConstructor{
-	"ipfs":  qipfs.NewFilesystem,
-	"local": localfs.NewFilesystem,
-	"http":  httpfs.NewFilesystem,
-	"mem":   qfs.NewMemFilesystem,
-	"map":   cafs.NewMapFilesystem,
+var constructors = map[string]qfs.Constructor{
+	qipfs.FilestoreType:   qipfs.NewFilesystem,
+	localfs.FilestoreType: localfs.NewFilesystem,
+	httpfs.FilestoreType:  httpfs.NewFilesystem,
+	qfs.MemFilestoreType:  qfs.NewMemFilesystem,
+	cafs.MapFilestoreType: cafs.NewMapFilesystem,
 }
 
 // New creates a new Mux Filesystem, if no Option funcs are provided,
 // New uses a default set of Option funcs. Any Option functions passed to this
 // function must check whether their fields are nil or not.
```

Comment:
Should probably make a note that the first added CAFS filesystem becomes the `defaultWriteDestination`